assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,643 @@
+            self.parse_token(tokens, &mut program.body)?;
+            end_of_nodes = program.body.len() == node_count;
+        }
+
+        // make sure at least one block has been read
+        if program.body.len() == beginning_node_count {
+            let start_op = tokens.read_at(start_pos).expect("no start token");
+            return Err(AssemblyError::empty_block(start_op));
+        }
+
+        // consume the 'end' token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_begin(
+                tokens.read_at(program_start).expect("no begin token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                Token::ELSE => Err(AssemblyError::dangling_else(token)),
+                _ => Err(AssemblyError::unmatched_begin(
+                    tokens.read_at(program_start).expect("no begin token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        Ok(())
+    }
+
+    // Parse procedure from token stream and add it to the procedure map in context.
+    fn parse_procedure(&mut self, tokens: &mut TokenStream) -> Result<(), AssemblyError> {
+        let proc_start = tokens.pos();
+
+        // read procedure name and consume the procedure header token
+        let header = tokens.read().expect("missing procedure header");
+        let (label, num_locals, is_export) = header.parse_proc()?;
+        tokens.advance();
+
+        let mut body = Vec::<Node>::new();
+
+        // parse procedure body
+        self.parse_token(tokens, &mut body)?;
+
+        // consume the 'end' token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_proc(
+                tokens.read_at(proc_start).expect("no proc token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                _ => Err(AssemblyError::unmatched_proc(
+                    tokens.read_at(proc_start).expect("no proc token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        // build and return the procedure
+        let proc = types::Procedure {
+            name: label,
+            num_locals,
+            is_export,
+            body,
+            index: self.procedures.len() as u32,
+        };
+        self.procedures.insert(proc.name.to_string(), proc);
+
+        Ok(())
+    }
+
+    // Parses a token from the token stream in a body, which generates a series of AST nodes.
+    fn parse_token(
+        &self,
+        tokens: &mut TokenStream,
+        nodes: &mut Vec<Node>,
+    ) -> Result<(), AssemblyError> {
+        if let Some(token) = tokens.read() {
+            match token.parts()[0] {
+                Token::ELSE => token.validate_else()?,
+                Token::IF => {
+                    token.validate_if()?;
+                    self.parse_if(tokens, nodes)?
```

Comment:
Once `parse_if` is updated to return a node, this can be changed to something like `node.push(self.parse_if(tokens, nodes)?);`.

Reply:
fixed